File: src/main/java/me/miki/shindo/management/mods/impl/CoordsMod.java
package me.miki.shindo.management.mods.impl;

import me.miki.shindo.Shindo;
import me.miki.shindo.management.event.EventTarget;
import me.miki.shindo.management.event.impl.EventRender2D;
import me.miki.shindo.management.language.TranslateText;
import me.miki.shindo.management.settings.config.Property;
import me.miki.shindo.management.settings.config.PropertyType;
import me.miki.shindo.management.mods.SimpleHUDMod;
import me.miki.shindo.management.settings.config.PropertyEnum;
import me.miki.shindo.management.nanovg.NanoVGManager;
import me.miki.shindo.management.nanovg.font.LegacyIcon;
import net.minecraft.util.BlockPos;
import net.minecraft.world.chunk.Chunk;

public class CoordsMod extends SimpleHUDMod {

    @Property(type = PropertyType.COMBO, translate = TranslateText.DESIGN)
    private Design design = Design.SIMPLE;
    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ICON, category = "Display")
    private boolean iconSetting = true;

    public CoordsMod() {
        super(TranslateText.COORDS, TranslateText.COORDS_DEDSCRIPTION, "coordinates");
    }

    @EventTarget
    public void onRender2D(EventRender2D event) {

        NanoVGManager nvg = Shindo.getInstance().getNanoVGManager();

        if (design == Design.SIMPLE) {
            this.draw();
        } else {
            nvg.setupAndDraw(() -> drawNanoVG());
        }
    }

    private void drawNanoVG() {

        String biome = "";
        Chunk chunk = mc.theWorld.getChunkFromBlockCoords(new BlockPos(mc.thePlayer));
        int maxWidth = 100;
        biome = chunk.getBiome(new BlockPos(mc.thePlayer), this.mc.theWorld.getWorldChunkManager()).biomeName;

        if (maxWidth < (this.getTextWidth("Biome: " + biome, 9, getHudFont(1)))) {
            maxWidth = (int) (this.getTextWidth("Biome: " + biome, 9, getHudFont(1)) + 12);
        } else {
            maxWidth = 107;
        }

        this.drawBackground(maxWidth, 48);
        this.drawText("X: " + (int) mc.thePlayer.posX, 5.5F, 5.5F, 9, getHudFont(1));
        this.drawText("Y: " + (int) mc.thePlayer.posY, 5.5F, 15.5F, 9, getHudFont(1));
        this.drawText("Z: " + (int) mc.thePlayer.posZ, 5.5F, 25.5F, 9, getHudFont(1));
        this.drawText("Biome: " + biome, 5.5F, 35.5F, 9, getHudFont(1));

        this.setWidth(maxWidth);
        this.setHeight(48);
    }

    @Override
    public String getText() {
        return "X: " + (int) mc.thePlayer.posX + " Y: " + (int) mc.thePlayer.posY + " Z: " + (int) mc.thePlayer.posZ + " ";
    }

    @Override
    public String getIcon() {
        return iconSetting ? LegacyIcon.MAP_PIN : null;
    }

    private enum Design implements PropertyEnum {
        SIMPLE(TranslateText.SIMPLE),
        FANCY(TranslateText.FANCY);

        private final TranslateText translate;

        Design(TranslateText translate) {
            this.translate = translate;
        }

        @Override
        public TranslateText getTranslate() {
            return translate;
        }
    }
}

File: src/main/kotlin/me/miki/shindo/management/mods/impl/CoordsMod.kt
package me.miki.shindo.management.mods.impl

import me.miki.shindo.Shindo.Companion.getInstance
import me.miki.client_api.event.EventTarget
import me.miki.shindo.management.event.impl.EventRender2D
import me.miki.shindo.management.language.TranslateText
import me.miki.shindo.management.mods.SimpleHUDMod
import me.miki.shindo.management.nanovg.font.LegacyIcon
import me.miki.shindo.management.settings.config.Property
import me.miki.shindo.management.settings.config.PropertyEnum
import me.miki.shindo.management.settings.config.PropertyType
import net.minecraft.util.BlockPos

class CoordsMod :
    SimpleHUDMod(TranslateText.COORDS, TranslateText.COORDS_DEDSCRIPTION, LegacyIcon.MOD_COORDS, "coordinates") {
    @Property(type = PropertyType.COMBO, translate = TranslateText.DESIGN)
    private val design = Design.SIMPLE

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ICON)
    private val iconSetting = true

    @EventTarget
    fun onRender2D(event: EventRender2D?) {
        val nvg = getInstance().nanoVGManager

        if (design == Design.SIMPLE) {
            this.draw()
        } else {
            nvg!!.setupAndDraw(Runnable { drawNanoVG() })
        }
    }

    private fun drawNanoVG() {
        var biome: String? = ""
        val chunk = mc.theWorld.getChunkFromBlockCoords(BlockPos(mc.thePlayer))
        var maxWidth = 100
        biome = chunk.getBiome(BlockPos(mc.thePlayer), this.mc.theWorld.worldChunkManager).biomeName

        if (maxWidth < (this.getTextWidth("Biome: " + biome, 9f, getHudFont(1))!!)) {
            maxWidth = (this.getTextWidth("Biome: " + biome, 9f, getHudFont(1))!! + 12).toInt()
        } else {
            maxWidth = 107
        }

        this.drawBackground(maxWidth.toFloat(), 48f)
        this.drawText("X: " + mc.thePlayer.posX.toInt(), 5.5f, 5.5f, 9f, getHudFont(1))
        this.drawText("Y: " + mc.thePlayer.posY.toInt(), 5.5f, 15.5f, 9f, getHudFont(1))
        this.drawText("Z: " + mc.thePlayer.posZ.toInt(), 5.5f, 25.5f, 9f, getHudFont(1))
        this.drawText("Biome: " + biome, 5.5f, 35.5f, 9f, getHudFont(1))

        this.setWidth(maxWidth)
        this.setHeight(48)
    }

    override fun getText(): String {
        return "X: " + mc.thePlayer.posX.toInt() + " Y: " + mc.thePlayer.posY.toInt() + " Z: " + mc.thePlayer.posZ.toInt() + " "
    }

    override fun getIcon(): String? {
        return if (iconSetting) LegacyIcon.MAP_PIN else null
    }

    private enum class Design(private val translate: TranslateText) : PropertyEnum {
        SIMPLE(TranslateText.SIMPLE),
        FANCY(TranslateText.FANCY);

        override fun getTranslate(): TranslateText {
            return translate
        }
    }
}


